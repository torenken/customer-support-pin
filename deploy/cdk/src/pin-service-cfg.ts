import { IValidation } from 'constructs';
import { z } from 'zod';

const accountSchema = z.string()
  .regex(/^\d+$/, { message: 'Amazon account number must be numeric' })
  .length(12, { message: 'Amazon account number be exactly 12 characters long' });

// ==============================================================================
// Stage Cfg

export declare type StageName = Stage.Dev | Stage.Prod

export enum Stage {
  Dev = 'dev',
  Prod = 'prod',
}

const stageCfgSchema = z.object({
  account: accountSchema,
});

export type StageCfg = z.infer<typeof stageCfgSchema>

export const validateStageCfg = (cfg: StageCfg): IValidation => {
  return new class implements IValidation {
    validate(): string[] {
      const messages: string[] = [];
      const result = stageCfgSchema.safeParse(cfg);
      if (!result.success) {
        messages.push(result.error.message);
      }
      return messages;
    }
  };
};

// ==============================================================================
// Pipeline Cfg

const pipelineCfgSchema = z.object({
  account: accountSchema,
  gitHubActionRoleArn: z.string(),
});

export type PipelineCfg = z.infer<typeof pipelineCfgSchema>

export const validatePipelineCfg = (cfg: PipelineCfg): IValidation => {
  return new class implements IValidation {
    validate(): string[] {
      const messages: string[] = [];
      const result = pipelineCfgSchema.safeParse(cfg);
      if (!result.success) {
        messages.push(result.error.message);
      }
      return messages;
    }
  };
};