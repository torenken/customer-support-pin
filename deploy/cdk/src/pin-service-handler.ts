import * as path from 'path';
import { GoFunction } from '@aws-cdk/aws-lambda-go-alpha';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export interface PinServiceHandlerProps{
  readonly serviceName: string;
  readonly environment: Record<string, string>;
}

export class PinServiceHandler extends GoFunction {
  constructor(scope: Construct, id: string, props: PinServiceHandlerProps) {
    super(scope, id, {
      entry: path.join(__dirname, `../../../api/services/${props.serviceName}`),
      functionName: `pin-service-${props.serviceName}`,

      memorySize: 1024,
      architecture: Architecture.ARM_64,
      runtime: Runtime.PROVIDED_AL2023,

      bundling: {
        goBuildFlags: ['-ldflags "-s -w"'],
        cgoEnabled: false,
      },
      environment: {
        ...props.environment,
      },
    });
  }
}
