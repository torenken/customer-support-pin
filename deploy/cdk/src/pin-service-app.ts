import { App, Tags } from 'aws-cdk-lib';
import { PipelineCfg, validatePipelineCfg, StageCfg, StageName } from './pin-service-cfg';
import { PinServicePipeline } from './pipeline/pin-service-pipeline';

const app = new App();

Tags.of(app).add('owner', 'customer-support-team');

const stageCfg = app.node.tryGetContext('stageConfig') as { [key in StageName]: StageCfg };
const pipelineCfg = app.node.tryGetContext('pipelineCfg') as PipelineCfg;

new PinServicePipeline(app, 'PinServicePipeline', {
  pipelineCfg,
  stageCfg,
  env: {
    account: pipelineCfg.account,
    region: 'eu-central-1',
  },
});

app.synth();

app.node.addValidation(validatePipelineCfg(pipelineCfg));
