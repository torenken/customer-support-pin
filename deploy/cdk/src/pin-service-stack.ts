import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { StageCfg } from './pin-service-cfg';
import { PinServiceHandler } from './pin-service-handler';

export interface PinServiceStackProps extends StackProps {
  readonly stageCfg: StageCfg;
}

export class PinServiceStack extends Stack {
  constructor(scope: Construct, id: string, props: PinServiceStackProps ) {
    super(scope, id, props);

    // define resources here...

    new PinServiceHandler(this, 'AddMoreHere', {
      serviceName: 'add-more-here',
      environment: {},
    });
  }
}
