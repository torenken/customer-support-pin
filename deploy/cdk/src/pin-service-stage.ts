import { Stage, StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { StageCfg, validateStageCfg } from './pin-service-cfg';
import { PinServiceStack } from './pin-service-stack';

export interface PinServiceStageProps extends StageProps {
  readonly stageCfg: StageCfg;
}

export class PinServiceStage extends Stage {
  constructor(scope: Construct, id: string, props: PinServiceStageProps) {
    super(scope, id, props);

    new PinServiceStack(this, 'PinService', {
      stageCfg: props.stageCfg,
    });

    this.node.addValidation(validateStageCfg(props.stageCfg));
  }
}
