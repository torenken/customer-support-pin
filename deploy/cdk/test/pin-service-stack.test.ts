import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { PinServiceStack } from '../src/pin-service-stack';

test('PinServiceStackSnapshot', () => {
  const app = new App({
    context: {
      'aws:cdk:bundling-stacks': ['NoStack'], //disable bundling lambda (asset), by using dummy stack-name (=> reduce the unit-test-time. jest-booster)
      '@aws-cdk/core:newStyleStackSynthesis': 'true',
    },
  });

  const stack = new PinServiceStack(app, 'PinServiceStack', {
    stageCfg: {
      account: '123456123456',
    },
  });

  const template = Template.fromStack(stack);

  //snapshot
  expect(template.toJSON()).toMatchSnapshot();
});
