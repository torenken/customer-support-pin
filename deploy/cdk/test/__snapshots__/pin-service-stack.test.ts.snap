// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinServiceStackSnapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AddMoreHere6413B8DE": {
      "DependsOn": [
        "AddMoreHereServiceRole317677DE",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "344e7879fe6261d42fa8eb24da2883999a6a7c6e08d61f68203cb1b5adbce53f.zip",
        },
        "FunctionName": "pin-service-add-more-here",
        "Handler": "bootstrap",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "AddMoreHereServiceRole317677DE",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AddMoreHereServiceRole317677DE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
